addons/mail/models/mail_activity.py

```diff
@@ -575,46 +575,132 @@ def activity_format(self):
             ]
         return activities
 
+    @api.model
+    def get_activity_data_format(self, activity_ids, message_ids=None):
+        completed_activities = self.env['mail.message'].browse(message_ids)._completed_activity_format() if message_ids else []
+        att_ids = list({attachment_id
+                        for activity in completed_activities
+                        for attachment_id in activity['attachment_ids']})
+        return {
+            'activities': self.env['mail.activity'].browse(activity_ids).activity_format(),
+            'attachments': self.env['ir.attachment'].browse(att_ids)._attachment_format() if att_ids else [],
+            'completed_activities': completed_activities,
+        }
+
     @api.model
     def get_activity_data(self, res_model, domain, limit=None, offset=0):
+        activity_types = self.env['mail.activity.type'].search(
+            ['|', ('res_model', '=', res_model), ('res_model', '=', False)])
+        display_done_activity_type_ids = [a.id for a in activity_types if a.display_done]
         activity_domain = [('res_model', '=', res_model)]
-        if domain:
-            res = self.env[res_model].search(domain)
+        domain_res_ids = self.env[res_model].search(domain).ids if domain else False
+        if domain or limit or offset:
+            res = self.env[res_model].search(domain or [], limit=limit, offset=offset)
```

Comment:
Isn't it possible to do one search and apply limit / filtering by hand on the returned ids ? That way we perform a single search.